src/nvim/api/window.c

```diff
@@ -583,3 +583,42 @@ Object nvim_win_call(Window window, LuaRef fun, Error *err)
   try_end(err);
   return res;
 }
+
+/// Gets the current viewport of a window
+///
+/// The returned tuple contains (everything is 0-based, end-exclusive):
+/// - Top-most buffer line index
+/// - Bottom-most buffer line index
+/// - Left-most column index
+/// - Right-most column index
+///
+/// @see |api-indexing|
+///
+/// @param window Window handle, or 0 for current window.
+/// @param[out] Error deailts, if any
+///
+/// @return (start_line, end_line, start_col, end_col) tuple of the viewport
+///         amount in this window.
+ArrayOf(Integer, 4) nvim_win_get_viewport(Window window, Error *err)
+  FUNC_API_SINCE(8)
+{
+  Array rv = ARRAY_DICT_INIT;
+  win_T *win = find_window_by_handle(window, err);
+
+  if (win) {
+    ADD(rv, INTEGER_OBJ(win->w_topline - 1));
+
+    linenr_T botline = win->w_botline;
+    if (buf_is_empty(win->w_buffer)) {
+      botline = 0;
+    } else if (win->w_empty_rows > 0) {
+      botline = win->w_buffer->b_ml.ml_line_count;
+    }
+    ADD(rv, INTEGER_OBJ(botline));
+
+    ADD(rv, INTEGER_OBJ(win->w_leftcol));
+    ADD(rv, INTEGER_OBJ(win->w_leftcol + win->w_width));
```

Comment:
This seems like it Mixes bytecols and screencols.

Reply:
Ouch indeed...

Reply:
Is this true? From what I can make out, `w_leftcol` is a screen column.

What would be a test for this? The following testcases pass:

```lua
    it('follows columns with multibyte chars', function()
      local get_viewport, win = prepare_win ({ "あいうえお" })
      meths.win_set_option(win, 'wrap', false)
      feed '1zl'
      eq({0, 1, 1, 5}, get_viewport())
    end)

    it('follows columns with tabs', function()
      local get_viewport, win, buf = prepare_win ({ "\ta\ta\ta\ta\ta\ta\ta\ta\ta\ta"})
      meths.win_set_option(win, 'wrap', false)
      meths.buf_set_option(buf, 'tabstop', 3)
      feed '1zl'
      eq({0, 1, 1, 5}, get_viewport())
    end)
```